File: examples/jCmpl/jCMPL-src/CmplMsg.java
/* ****************************************************************************
 * This code is part of jCMPL
 *
 * Copyright (C) 2013 Mike Steglich / B. Knie Technical University of Applied
 * Sciences Wildau, Germany
 *
 * pyCMPL is a project of the Technical University of Applied Sciences Wildau
 * and the Institute for Operations Research and Business Management at the
 * Martin Luther University Halle-Wittenberg.
 *
 * Please visit the project homepage <http://www.coliop.org>
 *
 * pyCMPL is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 3 of the License, or (at your option) any
 * later version.
 *
 * pyCMPL is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Lesser General Public # License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 * ****************************************************************************/

package jCMPL;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;
import org.apache.commons.lang3.StringEscapeUtils;

/**
 * Handling of a CMPL message for jCMPL
 * @author mike
 */
public class CmplMsg 
{
    
    private String _type;
    private String _file;
    private String _line;
    private String _description;
    
    /**
     * Constructor
     */
    public CmplMsg()    
    {
        _type = "";
        _file = "";
        _line = "";
        _description = "";
    } 

    /**
     * Returns the type of the messages
     * @return message type warning|error
     */
    public String type() {
        return _type;
    }

    /** 
     * Setter for the message type
     * Used by Cmpl
     * @param type message type
     */
    protected void setType(String type) {
        this._type = type;
    }

    /**
     * Returns the name of the CMPL file in that the error or warning occurs 
     * @return CMPL file name r CmplData file name
     */
    public String file() {
        return _file;
    }

    /**
     * Setter for the file name 
     * Used by CMPL 
     * @param file file name
     */
    protected void setFile(String file) {
        this._file = file;
    }

    /**
     * Returns the line in the CMPL file in that the error or warning occurs
     * @return line
     */
    public String line() {
        return _line;
    }

    /**
     * Setter fpr the line
     * Used by CMPL
     * @param line line
     */
    protected void setLine(String line) {
        this._line = line;
    }

    /**
     * Returns the a description of the error or warning message 
     * @return description of the error or warning  
     */
    public String description() {
        return _description;
    }

    protected void setDesrciption(String description) {
        this._description = description;
    }
}

/**
 * Handling of all CMPL messages for jCMPL
 * @author mike
 */
class CmplMessages {

    private String _cmplStatus;
    private String _cmplVersion;
    private String _cmplMessage;
    private String _msgFile;
    private ArrayList<CmplMsg> _cmplMessageList;

    /**
     * internal function - setting defaults
     */
    private void setDefaults() {
        _cmplStatus = "";
        _cmplVersion = "";
        _cmplMessage = "";
        _msgFile = "";
        _cmplMessageList = new ArrayList<CmplMsg>();
    }

    /**
     * Constructor
     */
    public CmplMessages() {
        setDefaults();
    }

   
    /**
     * Constructor using a CmplMessage file
     * @param msgFile 
     */
    public CmplMessages(String msgFile) {
        setDefaults();
        _msgFile = msgFile;
    }

    /**
     * Returns the CMPL related status of the Cmpl object
     * Used by Cmpl
     * @return cmpl status
     */
    public int cmplStatus() {
        int stat = -99;
        if (_cmplStatus.equalsIgnoreCase("normal")) {
            stat = Cmpl.CMPL_OK;
        } else if (_cmplStatus.equalsIgnoreCase("warning")) {
            stat = Cmpl.CMPL_WARNINGS;
        } else if (_cmplStatus.equalsIgnoreCase("error")) {
            stat = Cmpl.CMPL_FAILED;
        } else {
            stat = Cmpl.CMPL_UNKNOWN;
        }
        return stat;
    }

    /**
     * Returns a list of CmplMsg objects that contain the CMPL messages
     * @return List of CMPL messages
     */
    public ArrayList<CmplMsg> cmplMessageList() {
        return _cmplMessageList;
    }

    /* public String cmplMessage() {
        return _cmplMessage;
    }*/

    /**
     * Reads the CmplMessages into the jCMPL structures from th corresponding CmplMessage file
     * @throws CmplException 
     */
    protected void readCmplMessages() throws CmplException {
        readCmplMessages("");
    }

    /**
      * Reads the CmplMessages into the jCMPL structures from th corresponding CmplMessage file or a CmplMessage string 
     * @param msgStr CmplMessage string 
     * @throws CmplException 
     */
    protected void readCmplMessages(String msgStr) throws CmplException {

        if (msgStr.isEmpty() && _msgFile.isEmpty()) {
            throw new CmplException("Neither cmplMessageFile nor cmplMessageString defined");
        }

        ArrayList<String> lines = null;

        if (!_msgFile.isEmpty()) {
            String line = "";
            try {
                BufferedReader in = new BufferedReader(new FileReader(_msgFile));
                lines = new ArrayList<String>();

                while ((line = in.readLine()) != null) {
                    lines.add(line);
                }

                in.close();

            } catch (IOException err) {
                throw new CmplException("Cannot read message file :" + _msgFile + " ->" + err);
            }
        }  else {
            lines = new ArrayList<String>(Arrays.asList(msgStr.split("\n")));
        }

        int lineNr = 1;
        Boolean generalSection = false;
        Boolean msgSection = false;

        for (String line : lines) {

            if (lineNr == 1) {
                if (line.contains("<?xml version")) {
                    lineNr++;
                    continue;
                } else {
                    throw new CmplException("Cant't read cmplMessage file  - Not a XML file !");
                }
            }
            if (lineNr == 2) {
                if (line.contains("<CmplMessages")) {
                    lineNr++;
                    continue;
                } else {
                    throw new CmplException("Cant't read cmplMessage file  - Not a CmplMessages file !");
                }
            }
            if (line.contains("<general>")) {
                generalSection = true;
                msgSection = false;
                continue;
            }
            if (line.contains("</general>")) {
                generalSection = false;
                continue;
            }
            if (line.contains("<messages")) {
                generalSection = false;
                msgSection = true;
                continue;
            }
            if (line.contains("</messages")) {
                msgSection = false;
                continue;
            }
            if (generalSection) {
                if (line.contains("<generalStatus")) {
                    _cmplStatus = CmplTools.xmlStrToString(line, "<generalStatus>([^\"]*)</generalStatus>");
                    continue;
                }
                if (line.contains("<message")) {
                    _cmplMessage = CmplTools.xmlStrToString(line, "<message>([^\"]*)</message>");
                    continue;
                }
                if (line.contains("<cmplVersion")) {
                    _cmplVersion = CmplTools.xmlStrToString(line, "<cmplVersion>([^\"]*)</cmplVersion>");
                    continue;
                }
            }
            if (msgSection) {
                if (line.contains("<message")) {
                    ArrayList<String> tmpList = new ArrayList<String>();

                    String patternStr = "\"([^\"]*)\"";

                    Pattern pattern = Pattern.compile(patternStr);
                    Matcher matcher = pattern.matcher(line);

                    while (matcher.find()) {
                        tmpList.add(matcher.group(1));
                    }
                    CmplMsg x = new CmplMsg();
                    x.setType(tmpList.get(0));
                    x.setFile(tmpList.get(1));
                    x.setLine(tmpList.get(2));
                    x.setDesrciption(StringEscapeUtils.unescapeXml(tmpList.get(3)));
                    _cmplMessageList.add(x);

                }
            }
        }
    }

}

File: pyCmpl/lib/pyCmpl/CmplMsg.py
#***********************************************************************
 #  This code is part of pyCMPL 
 #
 #  Copyright (C) 2013
 #  Mike Steglich - Technical University of Applied Sciences
 #  Wildau, Germany 
 #
 #  pyCMPL is a project of the Technical University of 
 #  Applied Sciences Wildau and the Institute for Operations Research 
 #  and Business Management at the Martin Luther University 
 #  Halle-Wittenberg.
 #  Please visit the project homepage <www.coliop.org>
 # 
 #  pyCMPL is free software; you can redistribute it and/or modify it 
 #  under the terms of the GNU Lesser General Public License as published by 
 #  the Free Software Foundation; either version 3 of the License, or 
 #  (at your option) any later version.
 # 
 #  pyCMPL is distributed in the hope that it will be useful, but WITHOUT
 #  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 #  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 #  License for more details.
 #  
 #  You should have received a copy of the GNU Lesser General Public License
 #  along with this program; if not, see <http://www.gnu.org/licenses/>.
 #
 #**********************************************************************

#!/usr/bin/python 


import os
import re
import cStringIO 

from xml.sax.saxutils import unescape
from CmplDefs import *
from CmplException import *


#*************** CmplMsg *****************************************
class CmplMsg(object):
	def __init__(self):
		self.__type = ""
		self.__file = ""
		self.__line = ""
		self.__description = ""
		
	# getter and setter 	
	@property
	def type(self):
		return self.__type
	
	def setType(self, type):
		self.__type=type
		
	@property
	def file(self):
		return self.__file
		
	def setFile(self, file):
		self.__file=file
	
	@property
	def line(self):
		return self.__line
		
	def setLine(self, line):
		self.__line=line
		
	@property
	def description(self):
		return self.__description
		
	def setDescription(self, msg):
		self.__description=msg
	#end getter and setter
	
			
#*************** end CmplMsg *************************************


#*************** CmplMsg *****************************************
class CmplMessages(object):	

	def __init__(self, msgFile=None):
		self.__cmplStatus = ""
		self.__cmplVersion = ""
		self.__cmplMessage = ""
		self.__msgFile = msgFile
		self.__cmplMessageList = []
		
	#*********** cmplStatus ********	
	@property
	def cmplStatus(self):
		if self.__cmplStatus=="normal":
			return CMPL_OK
		elif self.__cmplStatus=="warning":
			return CMPL_WARNINGS
		elif self.__cmplStatus=="error":
			return CMPL_FAILED
		else:
			return CMPL_UNKNOWN
	#*********** end cmplStatus *****	
	
	#*********** cmplMessageList ****	
	@property
	def cmplMessageList(self):
		return self.__cmplMessageList
	#******* end cmplMessageList ****
	
	#*********** cmplMessage ********	
	@property
	def cmplMessage(self):
		return self.__cmplMessage
	#******* end cmplMessage ********
	
		
	#*********** readCmplMessages ********	
	def readCmplMessages(self, msgStr=None):
	
		if msgStr == None and self.__msgFile == None:
			raise CmplException("Neither cmplMessageFile nor cmplMessageString defined")
		
		if 	self.__msgFile != None:
			if not os.path.isfile(self.__msgFile):
				raise CmplException("CMPL failed."  )
	
			try:	
				f = open(self.__msgFile, "r")
				msgStr = f.read()
				f.close()
			except IOError, e:
				raise CmplServerException("IO error for instance file ")
				
				
		lines = cStringIO.StringIO(msgStr) 
		
		lineNr=1
		generalSection=False
		msgSection=False
		x = CmplMsg()
		
		for line in lines:
			if lineNr == 1:
				if not "<?xml version" in line: 
					raise CmplException("File " + self.__msgFile + " - is not a XML file!")
				else:
					lineNr += 1
					continue
			if lineNr == 2:
				if not "<CmplMessages" in line: 
					raise CmplException("Cant't read cmplMessage file " + self._cmplMsgFile + " - wrong file type!")
				else:
					lineNr += 1
					continue
	
			if "<general>" in line:
				generalSection=True
				msgSection=False
				continue
			if "</general>" in line:
				generalSection=False
				continue
			if "<messages" in line:
				generalSection=False
				msgSection=True
				continue
			if "</messages" in line:
				msgSection=False
				continue
				
			if generalSection:
				if "<generalStatus" in line:
					self.__cmplStatus = re.findall("<generalStatus>([^\"]*)</generalStatus>", line)[0]
					continue
				if "<message" in line:
					self.__cmplMessage = re.findall("<message>([^\"]*)</message>", line)[0]
					continue
				if "<cmplVersion" in line:
					self.__cmplVersion = re.findall("<cmplVersion>([^\"]*)</cmplVersion>", line)[0]
					continue
		
			if msgSection:
				if "<message" in line:
					tmpList=re.findall("\"([^\"]*)\"", line)
					x = CmplMsg()
					x.setType(tmpList[0])
					x.setFile(tmpList[1])
					x.setLine(tmpList[2])
					x.setDescription(unescape(tmpList[3]))
					self.__cmplMessageList.append(x)

						
							
	#*********** end readCmplMessages ****	
		
	

		
		
		
		

		